### PR TITLE
Update Terraform SDK to v2

### PR DESCRIPTION
**Pull request by dak1n1**
### Description

This PR updates the Terraform SDK v2, which drops support for Terraform 0.11.

This PR also includes a feature created by Ilia Lazebnik in PR #910, adding `allowed_topologies` to the storageclass resource and data source.

### Acceptance tests
- [x] Have you added an acceptance test for the functionality being added?
- [x] Have you run the acceptance tests on this branch?

### Release Note
Release note for [CHANGELOG](https://github.com/hashicorp/terraform-provider-kubernetes/blob/master/CHANGELOG.md):

```release-note
IMPROVEMENTS:

* Update Terraform SDK to v2 and drop support for Terraform 0.11.
* Add `allowed_topologies` to StorageClass (#391, #891)
```

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### CODE REVIEW

**Comment by dak1n1** [comment]
This isn't actually ready for review, but I'm marking it as such so I can run acceptance tests in TC.

**Comment by dak1n1** [comment]
Here are the results of acceptance tests run locally. https://gist.githubusercontent.com/dak1n1/d9a9e1c59a3ee959fc0185be3204fe5d/raw/7ae46bfdda01636f6abe1e8b94d37fd23720183d/gistfile1.txt

**Comment by dak1n1** [comment]
The test is running in GKE here. https://ci-oss.hashicorp.engineering/viewLog.html?buildId=153868

Looks like I have some work to do on some data sources.

File: kubernetes/resource_kubernetes_persistent_volume_claim_test.go [hunk 1]
```diff
@@ -36,45 +36,45 @@ func TestAccKubernetesPersistentVolumeClaim_basic(t *testing.T) {
...
-					return !isInMinikube, err
-				},
-				Check: resource.ComposeAggregateTestCheckFunc(
-					testAccCheckMetaAnnotations(&conf.ObjectMeta, map[string]string{
-						"TestAnnotationOne":                             "one",
-						"volume.beta.kubernetes.io/storage-provisioner": "k8s.io/minikube-hostpath",
-					}),
-				),
-			},
+			//			{ // GKE specific check
```
Comment (jrhouston) [inline kubernetes/resource_kubernetes_persistent_volume_claim_test.go:52]: Why'd we comment these ones out?

File: kubernetes/resource_kubernetes_service_test.go [hunk 2]
```diff
@@ -533,6 +524,94 @@ func TestAccKubernetesService_generatedName(t *testing.T) {
...
+							TargetPort: intstr.FromInt(80),
+						},
+					}),
+				),
+			},
+		},
+	})
+}
+
+func testAccCheckKubernetesServiceForceNew(old, new *api.Service, wantNew bool) resource.TestCheckFunc {
```
Comment (jrhouston) [inline kubernetes/resource_kubernetes_service_test.go:600]: If we're only checking the UID from the metadata I'd bet we could write a generic function for this using the dynamic client so we don't need to write this function for every resource. We can do that in another PR though.

File: kubernetes/resource_kubernetes_storage_class.go [hunk 3]
```diff
@@ -95,99 +128,124 @@ func resourceKubernetesStorageClassCreate(d *schema.ResourceData, meta interface
...
-	if storageClass.AllowVolumeExpansion != nil {
-		d.Set("allow_volume_expansion", *storageClass.AllowVolumeExpansion)
+		diags = append(diags, diag.FromErr(err)[0])
 	}
 
-	return nil
+	sc := flattenStorageClass(*storageClass)
+	for k, v := range sc {
+		err = d.Set(k, v)
+		log.Printf("DEBUGINFO: %v %v", k, v)
```
Comment (jrhouston) [inline kubernetes/resource_kubernetes_storage_class.go:179]: Leftover debug log?

File: kubernetes/resource_kubernetes_storage_class_test.go [hunk 4]
```diff
@@ -109,6 +182,33 @@ func TestAccKubernetesStorageClass_basic(t *testing.T) {
 					resource.TestCheckResourceAttr(resourceName, "allow_volume_expansion", "true"),
 					resource.TestCheckResourceAttr(resourceName, "mount_options.#", "0"),
 					resource.TestCheckResourceAttr(resourceName, "parameters.%", "0"),
+				),
+			},
+		},
+	})
+}
+
+func TestAccKubernetesStorageClass_allowedTopologies(t *testing.T) {
```
Comment (jrhouston) [inline kubernetes/resource_kubernetes_storage_class_test.go:191]:
```suggestion
func TestAccKubernetesStorageClass_allowedTopologies_minikube(t *testing.T) {
```

**Review by jrhouston — CHANGES_REQUESTED** [review]
Thanks for this heroic effort @dak1n1, this was **A LOT** of work. 💪 

I've looked at _every file_ other than the vendor directory. I see that most of it is just adding `Context` and wrapping errors into `Diagnostics`.

I've made a few comments – some questions and mostly cosmetic stuff.

File: GNUmakefile [hunk 5]
```diff
@@ -41,7 +42,13 @@ test: fmtcheck
 		xargs -t -n4 go test $(TESTARGS) -timeout=30s -parallel=4
 
 testacc: fmtcheck
-	TF_ACC=1 go test $(TEST) -v $(TESTARGS) -timeout 120m
+	rm -rf $(PWD)/.terraform || true
+	mkdir $(PWD)/.terraform
+	mkdir -p /tmp/.terraform.d/localhost/test/kubernetes/9.9.9/$(OS_ARCH) || true
```
[the remark this answers is not in the document]
Comment (dak1n1) [reply GNUmakefile:48]: yeah, this is the work-around that lets me test multiple versions of the provider. This way we'll catch any accidental ForceNew deletions during upgrade.

File: kubernetes/data_source_kubernetes_service_test.go [hunk 6]
```diff
@@ -41,11 +63,34 @@ func TestAccKubernetesDataSourceService_basic(t *testing.T) {
 }
 
 func testAccKubernetesDataSourceServiceConfig_basic(name string) string {
-	return testAccKubernetesServiceConfig_basic(name) + `
-data "kubernetes_service" "test" {
+	return fmt.Sprintf(`resource "kubernetes_service" "test" {
+	metadata {
```
[the remark this answers is not in the document]
Comment (dak1n1) [reply kubernetes/data_source_kubernetes_service_test.go]: Done... across 20 files even. ha

File: kubernetes/provider_test.go [hunk 7]
```diff
@@ -4,44 +4,62 @@ import (
...
-		"aws":        aws.Provider(),
-		"azurerm":    azurerm.Provider(),
+	}
+	testAccProviderFactories = map[string]func() (*schema.Provider, error){
+		"kubernetes": func() (*schema.Provider, error) {
+			return Provider(), nil
+		},
+	}
+	testAccExternalProviders = map[string]resource.ExternalProvider{
+		"kubernetes-local": {
```
[the remark this answers is not in the document]
Comment (dak1n1) [reply kubernetes/provider_test.go:39]: yeah, I think I can do a better job here now that I have gone through this process. This was the first solution I came up with, so it's a little rough around the edges, but I bet I can improve it. I'll give it a try and let you know how it goes.

[the remark this answers is not in the document]
Comment (dak1n1) [reply kubernetes/provider_test.go:39]: darn, I was excited to tidy this up. But I don't think it's possible. Here's the test I ran to cause a ForceNew:

First, I made this change without modifying anything else. This ensures that we can catch a ForceNew with the code as it is right now:

```
kubernetes/resource_kubernetes_pod.go
+       podSpecFields["dns_policy"].Default = "Default"
+       podSpecFields["dns_policy"].ForceNew = true

 TESTARGS="-run 'TestAccKubernetesPod_regression'" make testacc
```
This was the result, which is good! It caught the ForceNew:

```
=== RUN   TestAccKubernetesPod_regression
    resource_kubernetes_pod_test.go:968: Step 2/2 error: Check failed: 2 errors occurred:
                * Check 11/34 error: kubernetes_pod.test: Attribute 'spec.0.dns_policy' expected "ClusterFirst", got "Default"
                * Check 34/34 error: Expecting pod UIDs to be the same: expected 17675004-ae7d-4d07-8713-bf530ed49a6b got bcdea13f-0c9c-4568-b7bd-1a2ab166c719
--- FAIL: TestAccKubernetesPod_regression (61.76s)
FAIL
```

For the second test, I tidied up the code by removing the "kubernetes-local" provider. It was encouraging to see that it could run without that. I just passed "kubernetes" as the provider name instead:

```
kubernetes/resource_kubernetes_pod_test.go
@@ -968,6 +968,7 @@ func TestAccKubernetesPod_regression(t *testing.T) {                                                                                
        resource.Test(t, resource.TestCase{                                                                                                             
                PreCheck:          func() { testAccPreCheck(t) },                                                                                       
                ExternalProviders: testAccExternalProviders,
+               ProviderFactories: testAccProviderFactories,

-                               Config: requiredProviders() + testAccKubernetesPod_regression("kubernetes-local", name, imageName),
+                               Config: requiredProviders() + testAccKubernetesPod_regression("kubernetes", name, imageName),
```

However, the results were not as expected. The test did not catch the ForceNew. I'm not sure why. The full diff is [here](https://gist.githubusercontent.com/dak1n1/2e82808221842a56acd5e66b3e6fc593/raw/c0a47b868abf6d5181afbbaa9a88536f8bcd6d2a/gistfile1.txt), in case you want to look. 

```
=== RUN   TestAccKubernetesPod_regression
    resource_kubernetes_pod_test.go:968: Step 1/2 error: Check failed: 1 error occurred:
                * Check 11/33 error: kubernetes_pod.test: Attribute 'spec.0.dns_policy' expected "ClusterFirst", got "Default"
        
--- FAIL: TestAccKubernetesPod_regression (23.22s)
```

The reason I initially compiled the provider as an ExternalProvider is because I wasn't able to get the "terraform init" behavior I needed without it. The terraform init is needed in between TestSteps to use the newer version of the provider and replace the released version (simulating an upgrade). This is what I'll open an issue for upstream. Because I agree that my solution is really more of a work-around, and it would be nicer to have the SDK support this use case.

[the remark this answers is not in the document]
Comment (jrhouston) [reply kubernetes/provider_test.go:39]: Awesome, thanks for investigating. This is something I'd like to see in the SDK, so this work is good evidence for a case for it.

File: kubernetes/resource_kubernetes_ingress.go [hunk 8]
```diff
@@ -167,63 +166,77 @@ func resourceKubernetesIngressCreate(d *schema.ResourceData, meta interface{}) e
 		}
 
 		if len(res.Status.LoadBalancer.Ingress) > 0 {
-			return resource.NonRetryableError(resourceKubernetesIngressRead(d, meta))
+			diagnostics := resourceKubernetesIngressRead(ctx, d, meta)
+			if diagnostics.HasError() {
+				errmsg := diagnostics[0].Summary
+				return resource.NonRetryableError(fmt.Errorf("Error reading ingress: %v", errmsg))
+			}
+			return nil
```
[the remark this answers is not in the document]
Comment (dak1n1) [reply kubernetes/resource_kubernetes_ingress.go:174]: Good catch! Returning `nil` instead of `diag.Diagnostics` caused a crash once when I was migrating the helm provider... I don't think the two can be used interchangeably like `error` and `nil` can.

[the remark this answers is not in the document]
Comment (dak1n1) [reply kubernetes/resource_kubernetes_ingress.go:174]: oh oops, this part actually is a function within the function. So returning `nil` is correct. (This function returns a `*resource.RetryError`).

File: kubernetes/resource_kubernetes_persistent_volume_claim_test.go (hunk 1 above)
Comment (dak1n1) [reply kubernetes/resource_kubernetes_persistent_volume_claim_test.go]: This is because `testAccCheckMetaAnnotations` no longer works. I had to comment it out in all the test files. We can tackle that one as part of the Green Runs epic. Though I saw it as redundant... and really hard to debug. I gave it a try but didn't want to sink too much time into it.

File: kubernetes/resource_kubernetes_storage_class.go (hunk 3 above)
Comment (dak1n1) [reply kubernetes/resource_kubernetes_storage_class.go:179]: woops, thanks

File: kubernetes/resource_kubernetes_ingress.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (jrhouston) [reply kubernetes/resource_kubernetes_ingress.go:174]: Oh yeah, totally missed that thanks!

**Comment by dak1n1** [comment]
I was going to link the TC test I'm running, but I'm number 62 in the queue! Maybe it will show up here tomorrow. https://ci-oss.hashicorp.engineering/viewQueued.html?itemId=154938

**Comment by dak1n1** [comment]
@jrhouston I think this PR is ready to go. I fixed a few more tests since running TC here, so we're close to a green run. https://ci-oss.hashicorp.engineering/viewLog.html?buildId=155031&

File: kubernetes/resource_kubernetes_affinity_test.go [hunk 9]
```diff
@@ -183,31 +188,32 @@ func TestAccKubernetesPod_with_pod_anti_affinity_with_preferred_during_schedulin
...
+					resource.TestCheckResourceAttr("kubernetes_pod.test", fmt.Sprintf("%s.0.pod_affinity_term.0.label_selector.0.match_expressions.#", keyName), "1"),
+					resource.TestCheckResourceAttr("kubernetes_pod.test", fmt.Sprintf("%s.0.pod_affinity_term.0.label_selector.0.match_expressions.0.key", keyName), "security"),
+					resource.TestCheckResourceAttr("kubernetes_pod.test", fmt.Sprintf("%s.0.pod_affinity_term.0.label_selector.0.match_expressions.0.operator", keyName), "NotIn"),
+					resource.TestCheckResourceAttr("kubernetes_pod.test", fmt.Sprintf("%s.0.pod_affinity_term.0.label_selector.0.match_expressions.0.values.#", keyName), "2"),
+					resource.TestCheckResourceAttr("kubernetes_pod.test", fmt.Sprintf("%s.0.pod_affinity_term.0.label_selector.0.match_expressions.0.values.0", keyName), "bar"),
+					resource.TestCheckResourceAttr("kubernetes_pod.test", fmt.Sprintf("%s.0.pod_affinity_term.0.label_selector.0.match_expressions.0.values.0", keyName), "foo"),
+					resource.TestCheckResourceAttr("kubernetes_pod.test", fmt.Sprintf("%s.0.pod_affinity_term.0.label_selector.0.match_labels.%%", keyName), "0"),
```
[the remark this answers is not in the document]
Comment (dak1n1) [reply kubernetes/resource_kubernetes_affinity_test.go:213]: This is an escape character for `fmt.Sprintf`: `%`. I had to use it so I could literally print out `%` here like the other tests do. So that's why there are two of them.

**Review by alexsomesan — COMMENTED** [review]
Wow, this was really quite an undertaking! Thanks a million for doing all of this.

I'm slowly going through all the changes.
I thought I'd leave comments in batches since it might be faster to address them while I keep reading through the PR.

File: .terraformrc [hunk 10]
```diff
@@ -0,0 +1,10 @@
+provider_installation {
```
[the remark this answers is not in the document]
Comment (dak1n1) [reply .terraformrc:1]: Good idea! I do have a directory we can use for this: `kubernetes/test-infra/external-providers`. I'll use that.

File: GNUmakefile [hunk 11]
```diff
@@ -41,7 +43,13 @@ test: fmtcheck
...
```
[the remark this answers is not in the document]
Comment (dak1n1) [reply GNUmakefile:48]: Yeah, I really wanted to use a different directory. The problem I ran into is that during the terraform apply (inside the acceptance test), it was unable to find the relative path I gave it. So I had to choose a directory that I knew would be there on most systems (/tmp). I can experiment with others though and let you know.

File: kubernetes/data_source_kubernetes_ingress.go [hunk 12]
```diff
@@ -15,15 +17,14 @@ func dataSourceKubernetesIngress() *schema.Resource {
...
-		Read: dataSourceKubernetesIngressRead,
+		ReadContext: dataSourceKubernetesIngressRead,
 
 		Schema: map[string]*schema.Schema{
 			"metadata": namespacedMetadataSchema("ingress", false),
 			"spec": {
 				Type:        schema.TypeList,
 				Description: docIngress["spec"],
 				Computed:    true,
-				MaxItems:    1,
```
[the remark this answers is not in the document]
Comment (dak1n1) [reply kubernetes/data_source_kubernetes_ingress.go:26]: The new SDK version doesn't allow using MaxItems together with Computed. It throws this error:

```
MaxItems is for configurable attributes,there's nothing to configure on computed-only field
```

File: kubernetes/data_source_kubernetes_persistent_volume_claim_test.go [hunk 13]
```diff
@@ -4,40 +4,47 @@ import (
...
 	name := fmt.Sprintf("tf-acc-test-%s", acctest.RandStringFromCharSet(10, acctest.CharSetAlphaNum))
 
 	resource.Test(t, resource.TestCase{
-		PreCheck:  func() { testAccPreCheck(t) },
-		Providers: testAccProviders,
+		PreCheck:          func() { testAccPreCheck(t) },
+		ProviderFactories: testAccProviderFactories,
 		Steps: []resource.TestStep{
-			{
+			{ // The first apply creates the resource. The second apply reads the resource using the data source.
```
[the remark this answers is not in the document]
Comment (dak1n1) [reply kubernetes/data_source_kubernetes_persistent_volume_claim_test.go:18]: Thanks. Yeah, there might actually be a regression... because these used to work in one apply. But now they require two applies. But when I looked up data sources in the Terraform docs and related issues, this seems to be the expected behavior.

Most data sources will be evaluated before the resource it depends on is evaluated. The exception is when the attribute it's fetching is computed. Only then will it wait for the resource to be created before evaluating the data source.

```
Terraform only delays reading a data resource if its configuration includes a value that can't be determined until the apply step.

The pattern of reading something using a data resource in the same configuration where it's created does tend to lead to this sort of problem, and so we'd recommend avoiding that where possible. 
```

https://github.com/hashicorp/terraform/issues/23242#issuecomment-548574212

Most times, users should reference values from the resource itself, instead of using a data source. So I'm hoping we're setting the right example by doing this in two applies.

File: kubernetes/data_source_kubernetes_persistent_volume_claim_test.go [hunk 14]
```diff
@@ -80,17 +87,15 @@ func testAccKubernetesDataSourcePVCConfig_basic(name string) string {
   }
 
   wait_until_bound = false
-  lifecycle {
```
[the remark this answers is not in the document]
Comment (dak1n1) [reply kubernetes/data_source_kubernetes_persistent_volume_claim_test.go:83]: Yeah, I removed this to get the test to pass. Here's what I have in my notes about it:

```
Error: Computed attribute cannot be set
on config992760292/terraform_plugin_test.tf line 2, in resource "kubernetes_persistent_volume_claim" "test":
metadata {

# solution: remove the ignore_changes block
   wait_until_bound = false
-  lifecycle {
-    ignore_changes = [metadata]
-  }
```

File: GNUmakefile (hunk 11 above)
[the remark this answers is not in the document]
Comment (dak1n1) [reply GNUmakefile:48]: I see why I didn't do this. It's because I would need to hard-code the path in this file: .terraformrc

```
provider_installation {
  filesystem_mirror {
    path    = "/tmp/.terraform.d"
...
```

This is the file that acceptance tests read, and it doesn't seem to work with a relative path. So it looks like we're stuck with /tmp for now.

**Comment by dak1n1** [comment]
I updated the TC test to include the new .terraformrc path. It looks like it's running well now. https://ci-oss.hashicorp.engineering/viewLog.html?buildId=155606&

**Comment by dak1n1** [comment]
That's better:

```
$ TESTARGS="-run 'googleCloud'" make testacc
=== RUN   TestAccKubernetesPersistentVolumeClaim_googleCloud_volumeMatch
--- PASS: TestAccKubernetesPersistentVolumeClaim_googleCloud_volumeMatch (148.40s)
=== RUN   TestAccKubernetesPersistentVolumeClaim_googleCloud_volumeUpdate
--- PASS: TestAccKubernetesPersistentVolumeClaim_googleCloud_volumeUpdate (76.47s)
=== RUN   TestAccKubernetesPersistentVolumeClaim_googleCloud_storageClass
--- PASS: TestAccKubernetesPersistentVolumeClaim_googleCloud_storageClass (13.65s)
=== RUN   TestAccKubernetesPersistentVolume_googleCloud_basic
--- PASS: TestAccKubernetesPersistentVolume_googleCloud_basic (80.44s)
=== RUN   TestAccKubernetesPersistentVolume_googleCloud_volumeSource
--- PASS: TestAccKubernetesPersistentVolume_googleCloud_volumeSource (72.04s)
PASS
ok      github.com/hashicorp/terraform-provider-kubernetes/kubernetes   391.045s
```

File: GNUmakefile (hunk 11 above)
[the remark this answers is not in the document]
Comment (alexsomesan) [reply GNUmakefile:49]: Alright, sounds like this is the safest choice for now. If it runs well in TC, then let's go with it.

File: kubernetes/data_source_kubernetes_persistent_volume_claim_test.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (alexsomesan) [reply kubernetes/data_source_kubernetes_persistent_volume_claim_test.go:18]: TBH, i wasn't aware of that nuance in the in evaluation of datasources and computed attributes. Thanks for uncovering that bit of info. It does make a lot of sense when that's the documented expected behaviour.

File: kubernetes/resource_kubernetes_persistent_volume_test.go [hunk 15]
```diff
@@ -184,6 +184,7 @@ func TestAccKubernetesPersistentVolume_googleCloud_basic(t *testing.T) {
 					resource.TestCheckResourceAttr("kubernetes_persistent_volume.test", "spec.0.persistent_volume_source.0.gce_persistent_disk.#", "1"),
 					resource.TestCheckResourceAttr("kubernetes_persistent_volume.test", "spec.0.persistent_volume_source.0.gce_persistent_disk.0.pd_name", diskName),
 				),
+				ExpectNonEmptyPlan: true, // GCP adds label "goog-gke-volume" to the disk.
```
Comment (alexsomesan) [inline kubernetes/resource_kubernetes_persistent_volume_test.go:187]: I wonder if doing this will mask a whole set of potential issues as we will be ignoring any unexpected diffs on the whole resource.
Regardless, I think this PR is already doing a lot, so it would best to continue this conversation in a different context.

**Review by alexsomesan — APPROVED** [review]
Again, thanks a lot for the effort of seeing this through. It's a massive chunk of work!

I've checked this branch in a few of the CI stages and it looks like it does very similarly to master in terms of flaky tests.

I think we can go ahead and merge and address the test issues individually, like usual.

File: kubernetes/resource_kubernetes_persistent_volume_test.go (hunk 15 above)
Comment (dak1n1) [reply kubernetes/resource_kubernetes_persistent_volume_test.go:187]: That's a good point. This kind of worries me too. It's a label placed the google disk resource (a resource owned by another provider). I could probably use something like this:

```
lifecycle {
    ignore_changes = [labels]
  }
```

I might give it a test in another PR.

**Comment by dak1n1** [comment]
Thanks for the thorough reviews on this beast of a PR! @alexsomesan @jrhouston :partying_face:
